Improve sensor simulation

andLaing opened the pull request:
fee.NOISE_DAQ reduced to remove double counting
in noise stages of PMT electronics simulation.

fee.ADC_TO_PES recalculated using faked calibration
pulses and changed to 18.71 to better calculate the
gain dispersion of the PMTs.

In sensor_functions:

Gain dispersion application changed position so that
it doesn't affect the noise which was found not to match
dispersion in noise seen in data.

charge_fluctuation function added to simulate the single
pe fluctuation using the RMS values in the database.

paolafer committed:
Improved sensor simulation

fee.NOISE_DAQ reduced to remove double counting
in noise stages of PMT electronics simulation.

fee.ADC_TO_PES recalculated using faked calibration
pulses and changed to 18.71 to better calculate the
gain dispersion of the PMTs.

In sensor_functions:

Gain dispersion application changed position so that
it doesn't affect the noise which was found not to match
dispersion in noise seen in data.

charge_fluctuation function added to simulate the single
pe fluctuation using the RMS values in the database.

gonzaponte commented on invisible_cities/cities/base_cities.py:
```diff
@@ -259,16 +259,32 @@ def set_up_database(self):
+                                           out=np.zeros(len(self.all_pmt_adc_to_pes)),
+                                           where=self.all_pmt_adc_to_pes!=0          )
+        sipm_single_pe_rms     = DataSiPM.Sigma.values         .astype(np.double)
+        if not sipm_single_pe_rms.any():
```
Shouldn't this be handled by the database?

andLaing replied:
In principle it could be, I'll look into it but it might imply more changes than this implementation

gonzaponte commented on invisible_cities/cities/base_cities.py:
```diff
@@ -259,16 +259,32 @@ def set_up_database(self):
+        pmt_single_pe_rms      = DataPMT.Sigma.values          .astype(np.double)
+        self.pmt_pe_resolution = np.divide(pmt_single_pe_rms                         ,
+                                           self.all_pmt_adc_to_pes                   ,
+                                           out=np.zeros(len(self.all_pmt_adc_to_pes)),
```
`np.zeros_like(self.all_pmt_adc_to_pes)`

andLaing replied:
I tried this in my initial testing in interactive but it was giving me an error. I'll try again directly in the code.

gonzaponte commented on invisible_cities/cities/base_cities.py:
```diff
@@ -259,16 +259,32 @@ def set_up_database(self):
+        self.pmt_pe_resolution = np.divide(pmt_single_pe_rms                         ,
+                                           self.all_pmt_adc_to_pes                   ,
+                                           out=np.zeros(len(self.all_pmt_adc_to_pes)),
+                                           where=self.all_pmt_adc_to_pes!=0          )
```
spaces around `!=`

gonzaponte commented on invisible_cities/cities/base_cities.py:
```diff
@@ -259,16 +259,32 @@ def set_up_database(self):
+                                            self.sipm_adc_to_pes                   ,
+                                            out=np.zeros(len(self.sipm_adc_to_pes)),
+                                            where=self.sipm_adc_to_pes!=0          )
 
```
`zeros_like` and spaces around `!=`

gonzaponte replied:
also align equal signs and surround them with spaces

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -7,11 +7,28 @@
+    if single_pe_rms == 0:
+        ## Protection for some versions of numpy etc
+        return signal
+    else:
```
`else` and indentation not needed

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -7,11 +7,28 @@
+    else:
+        ## We need to convert to float to get accuracy here
+        sig_fl = signal.astype(float)
+        sigma = np.sqrt(sig_fl[sig_fl>0]) * single_pe_rms
```
assign `sig_fl > 0` to a variable and surround the `>` and `<` signs with spaces.

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -26,30 +43,42 @@ def simulate_pmt_response(event, pmtrd, adc_to_pes, run_number = 0):
     # FEE, with noise PMT
     fee  = FE.FEE(run_number,
                   noise_FEEPMB_rms=FE.NOISE_I, noise_DAQ_rms=FE.NOISE_DAQ)
+    #
```
Empty comment? Use blank lines to separate code blocks.

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -26,30 +43,42 @@ def simulate_pmt_response(event, pmtrd, adc_to_pes, run_number = 0):
-    return wfm.to_adc(sipmrd[event], sipm_adc_to_pes) + sipms_noise_sampler.sample()
+
+    ## Fluctuate according to charge resolution
+    sipm_fl = np.array([charge_fluctuation(si, rms) for si, rms in zip(sipmrd[event], pe_resolution)])
```
may I suggest `sipm_fl = np.array(tuple(map(charge_fluctuation, sipmrd[event], pe_resolution)))`

gonzaponte commented on invisible_cities/sierpe/fee.py:
```diff
@@ -29,7 +29,7 @@
 f_LPF2 = 10 * units.MHZ
 ADC_TO_PES_LPF = 24.1  # After LPF, comes out from spe area
-ADC_TO_PES     = 23.1
+ADC_TO_PES     = 18.71
```
I guess the consequences of changing these numbers are controled, right?

andLaing replied:
I'd add the plots if I could (and I could find them at the moment) but the changes were checked by removing the variation and checking the first non pedestal peak before the 1pe fluctuation was added. 18.71 is the average position over the 12 pmts. Adding the dispersion after this change gives positions consistent with the DB gains.

gonzaponte commented on invisible_cities/sierpe/fee.py:
```diff
@@ -129,13 +129,13 @@ def spe_pulse_train(spe,
 
-    spe_pulse = signal.convolve(cnt[0:-len(spe.spe)+1], spe.spe)
+    spe_pulse = signal.convolve(cnt[0:-len(spe.spe)+1], spe.spe * norm)
     return spe_pulse
```
the zero before the colon can be removed. surround the `+` sign with spaces.

gonzaponte commented on invisible_cities/sierpe/fee.py:
```diff
@@ -173,7 +173,7 @@ def __init__(self, run_number = 0, gain=FEE_GAIN,
 
         DataPMT = DB.DataPMT(run_number)
-
+        
```
remove this whitespaces, please

gonzaponte commented on invisible_cities/cities/base_cities.py:
```diff
@@ -259,16 +259,32 @@ def set_up_database(self):
+        self.coeff_c             = DataPMT.coeff_c.values        .astype(np.double)
+        self.coeff_blr           = DataPMT.coeff_blr.values      .astype(np.double)
+        self.noise_rms           = DataPMT.noise_rms.values      .astype(np.double)
+
```
you have introduced a double space before the `=` signs

gonzaponte reviewed (commented):
Some minor comments to uniform style and some questions about the implications of the changes introduced.

andLaing committed:
Implementation of reviewers requests

Tidied up pythonification

Moved check of zeros in SiPM database 1pe Sigmas
to the database function DataSiPM.

gonzaponte commented on invisible_cities/cities/base_cities.py:
```diff
@@ -270,6 +270,18 @@ def set_up_database(self):
+                                            self.sipm_adc_to_pes                   ,
+                                            out=np.zeros_like(self.sipm_adc_to_pes),
+                                            where=self.sipm_adc_to_pes != 0        )
+
```
align the `=` in `out` and `where` and surround them with spaces in both `np.divide` calls

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -7,11 +7,29 @@
 def convert_channel_id_to_IC_id(data_frame, channel_ids):
     return pd.Index(data_frame.ChannelID).get_indexer(channel_ids)
 
+def charge_fluctuation(signal, single_pe_rms):
```
use double blank lines to separate function definitions

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -7,11 +7,29 @@
+    ## We need to convert to float to get accuracy here
+    sig_fl = signal.astype(float)
+    non_zero = sig_fl > 0
+    sigma = np.sqrt(sig_fl[non_zero]) * single_pe_rms
```
align `=`

gonzaponte commented on invisible_cities/reco/sensor_functions.py:
```diff
@@ -7,11 +7,29 @@
+    sigma = np.sqrt(sig_fl[non_zero]) * single_pe_rms
+    sig_fl[non_zero] = np.random.normal(sig_fl[non_zero], sigma)
+    ## This fluctuation can't give negative signal
+    sig_fl[sig_fl<0] = 0
```
surround `<` with spaces

gonzaponte reviewed (commented):
Just a couple of cosmetic requests before approval

Commit:
Cosmetics

gonzaponte reviewed (approved):
This PR adds charge fluctuation to the simulation of the sensors' response. The database functions have been modified to provide the relevant values for this task. Finally, some fee variables have been updated.
